### PR TITLE
fix: optimize db queries

### PR DESCRIPTION
**Pull request by morganridel**
~PR is not perfectly ready yet, I will try to remove the last bit of formatting problems tomorrow~ Should be better now

It is not a good idea to do x*n DB request when receiving n ids as input, I tried to reduce to a constant x queries (independent on input size) for the item-memberships call while keeping the same behavior.

Also GIST indices on DB work for all the ltree operators (`@>` ...). They were tested with `EXPLAIN ANALYZE` on realistic queries directly on the DB and are retrieving the ancestors instantly.

Also fix `ItemTagRepository.hasManyForMany` which didn't work

Depending on results on performance I should take a look at the following in the future:

* Optimizing `/children` call queries (should be better with the new index)
*  Optimizing queries from the `ItemTag` repository
* Websockets is a pain as it do a `validatePermission` for EACH item channel on app load, sending a lot of DB requests

### CODE REVIEW

File: src/services/item/service.ts [hunk 1]
```diff
@@ -112,17 +112,21 @@ export class ItemService {
...
+      if (memberships?.data[id] === undefined) {
         delete result.data[id];
-        if (e instanceof CoreError || e instanceof Error) {
-          result.errors.push(e);
-        }
       }
     }
-    return result;
+
+    return { data: result.data, errors: [...(memberships?.errors ?? []), ...result.errors] };
```
Comment (pyphilia) [inline src/services/item/service.ts:129]:
```suggestion
    return { data: result.data, errors: result.errors.concat(memberships?.errors ?? []) };
```

File: src/services/itemMembership/index.ts [hunk 2]
```diff
@@ -36,7 +36,7 @@ const plugin: FastifyPluginAsync = async (fastify) => {
       fastify.get<{ Querystring: { itemId: string[] } }>(
         '/',
         { schema: getItems, preHandler: fastify.fetchMemberInSession },
-        async ({ member, query: { itemId: ids }, log }) => {
+        async ({ member, id, query: { itemId: ids }, log }) => {
```
Comment (pyphilia) [inline src/services/itemMembership/index.ts:39]: `id`?

File: src/services/itemMembership/repository.ts [hunk 3]
```diff
@@ -97,26 +97,22 @@ export const ItemMembershipRepository = AppDataSource.getRepository(ItemMembersh
...
+    const ids = items.map((i) => i.id);
+    const query = this.createQueryBuilder('item_membership')
+      .innerJoin('item', 'descendant', 'item_membership.item_path @> descendant.path')
+      .where('descendant.id in (:...ids)', { ids: ids });
     if (memberId) {
       query.andWhere('member.id = :memberId', { memberId });
     }
 
+    query
+      .leftJoinAndSelect('item_membership.item', 'item')
```
Comment (pyphilia) [inline src/services/itemMembership/repository.ts:113]: Could you exchange `innerJoin` to `innerJoinAndSelect` and remove this line?

File: src/services/itemMembership/repository.ts [hunk 4]
```diff
@@ -133,6 +129,54 @@ export const ItemMembershipRepository = AppDataSource.getRepository(ItemMembersh
...
+    }
+
+    const ids = items.map((i) => i.id);
+
+    const query = ItemMembershipRepository.createQueryBuilder('item_membership')
+      // Map each membership to the item it can affect
+      .innerJoin('item', 'descendant', 'item_membership.item_path @> descendant.path')
+      // Join for entity result
+      .leftJoinAndSelect('item_membership.member', 'member')
+      .leftJoinAndSelect('item_membership.item', 'item')
```
Comment (pyphilia) [inline src/services/itemMembership/repository.ts:148]: Select from the innerjoin ?

File: src/services/utils.ts [hunk 5]
```diff
@@ -53,3 +54,33 @@ export const buildItemLink = (item: Item) => {
...
+  // eslint-disable-next-line prefer-const
+  let [sql, params] = query.getQueryAndParameters();
+
+  params.forEach((value, index) => {
+    if (typeof value === 'string') {
+      sql = sql.replace(`$${index + 1}`, `'${value}'`);
+    }
+    if (typeof value === 'object') {
+      if (Array.isArray(value)) {
+        sql = sql.replace(
```
Comment (pyphilia) [inline src/services/utils.ts:70]: Maybe create another file for this debug function.

**Review by pyphilia — APPROVED** [review]
It looks really good, I'm gonna test it locally and try to deploy on dev 🍔

File: src/services/itemMembership/repository.ts (hunk 3 above)
Comment (morganridel) [reply src/services/itemMembership/repository.ts:113]: No this is not the same join, the first one is just used as a filter to retrieve memberships that are parents, but this is not a join where `item_membership.item_path = item.path`. The second one is for filling the entity that will be returned with the real linked item
I.E: I need to do the two join, at least from what I was trying to do

**Comment by pyphilia** [comment]
I've added a bunch of tests for `validatePermission` and `validatePermissionMany` (I've caught a small bug). I think I've covered most of the cases.

**Comment by morganridel** [comment]
Really great idea to add unit tests, perfect usecase :)

File: src/services/authorization.test.ts [hunk 6]
```diff
@@ -0,0 +1,2660 @@
+import { ItemTagType, PermissionLevel, UUID } from '@graasp/sdk';
```
Comment (dialexo) [inline src/services/authorization.test.ts:1]: Very nice all those unit tests :sparkles: for readability purposes, i would suggest adding blank newlines between `it` blocks as well as `describe` blocks. It makes it easier to distinguish the test cases visually.

File: src/services/authorization.ts [hunk 7]
```diff
@@ -26,27 +28,120 @@ const permissionMapping = {
...
+  ]);
+
+  const result: ResultOf<ItemMembership | null> = {
+    data: inheritedMemberships?.data ?? {},
+    errors: [],
+  };
+
+  for (const item of items) {
+    const highest = result.data[item.id]?.permission;
+    const isValid = highest && permissionMapping[highest].includes(permission);
```
Comment (dialexo) [inline src/services/authorization.ts:60]: Is this equivalent to (i.e. can you replace instances of `permissionMapping` with):
```suggestion
    const isValid = highest && PermissionLevelCompare.lte(permission, highest);
```

File: src/services/authorization.ts [hunk 8]
```diff
@@ -26,27 +28,120 @@ const permissionMapping = {
...
+    const isPublic = tags.data[item.id].includes(ItemTagType.Public);
+
+    // HIDDEN CHECK - prevent read
+    // cannot read if your have read access only
+    if (highest === PermissionLevel.Read || (isPublic && !highest)) {
```
Comment (dialexo) [inline src/services/authorization.ts:65]: Q: can members with write access view hidden items?

File: src/services/authorization.ts [hunk 9]
```diff
@@ -26,27 +28,120 @@ const permissionMapping = {
...
   item: Item,
 ): Promise<ItemMembership | null> => {
   // get best permission for user
   // but do not fetch membership for signed out member
+
   const inheritedMembership = member
     ? await itemMembershipRepository.getInherited(item, member, true)
     : null;
   const highest = inheritedMembership?.permission;
   const isValid = highest && permissionMapping[highest].includes(permission);
```
Comment (dialexo) [inline src/services/authorization.ts:133]: Same as above

File: src/services/item/plugins/itemTag/repository.ts [hunk 10]
```diff
@@ -60,14 +60,22 @@ export const ItemTagRepository = AppDataSource.getRepository(ItemTag).extend({
       }),
     );
 
-    const hasTags = await query
+    const hasTags: ItemTag[] = await query
       .andWhere('itemTag.type IN (:...types)', { types: tagTypes })
-      .getOne();
+      .getMany();
```
Comment (dialexo) [inline src/services/item/plugins/itemTag/repository.ts:65]: Info: why did the previous implementation not return many?

File: src/services/itemMembership/repository.ts [hunk 11]
```diff
@@ -97,26 +97,22 @@ export const ItemMembershipRepository = AppDataSource.getRepository(ItemMembersh
...
-      }),
-    );
+    if (items.length === 0) {
+      return { data: {}, errors: [] };
+    }
 
+    const ids = items.map((i) => i.id);
+    const query = this.createQueryBuilder('item_membership')
+      .innerJoin('item', 'descendant', 'item_membership.item_path @> descendant.path')
+      .where('descendant.id in (:...ids)', { ids: ids });
```
Comment (dialexo) [inline src/services/itemMembership/repository.ts:107]:
```suggestion
      .where('descendant.id in (:...ids)', { ids });
```

File: src/services/itemMembership/repository.ts [hunk 12]
```diff
@@ -133,6 +129,54 @@ export const ItemMembershipRepository = AppDataSource.getRepository(ItemMembersh
...
+      query.andWhere('item.id not in (:...ids)', { ids: ids });
+    }
+    // Keep only closest membership per descendant
+    query
+      .addSelect('descendant.id')
+      .distinctOn(['descendant.id'])
+      .orderBy('descendant.id')
+      .addOrderBy('nlevel(item_membership.item_path)', 'DESC');
+
+    // annoyingly, getMany removes dupplicate entities, however in this case two items might be linked to the same effective membership
```
Comment (dialexo) [inline src/services/itemMembership/repository.ts:162]:
```suggestion
    // annoyingly, getMany removes duplicate entities, however in this case two items might be linked to the same effective membership
```

File: src/services/utils.ts [hunk 13]
```diff
@@ -53,3 +54,33 @@ export const buildItemLink = (item: Item) => {
 
   return itemLink;
 };
+
+// Print a query with parameters already filled in, ready to be sent to the database
+// Only for debug purposes. DO NOT USE in production. (injection risk)
+export const printFilledSQL = <T extends ObjectLiteral>(query: QueryBuilder<T>) => {
```
Comment (dialexo) [inline src/services/utils.ts:60]: Consider running a `if (NODE_ENV === 'production') { throw new Error() }` here?

**Review by dialexo — APPROVED** [review]
Wonderful work :strawberry: I've left a few questions and minor improvements :+1:

**Review by spaenleh — APPROVED** [review]
Very nice ! Nice tests too !

File: src/services/item/plugins/itemTag/repository.ts (hunk 10 above)
Comment (morganridel) [reply src/services/item/plugins/itemTag/repository.ts:65]: The previous implementation didn't work and wasn't used

File: src/services/authorization.ts (hunk 8 above)
Comment (morganridel) [reply src/services/authorization.ts:65]: I guess so, so you can change the hidden status? I just kept the existing behavior

File: src/services/authorization.ts (hunk 7 above)
Comment (morganridel) [reply src/services/authorization.ts:60]: I don't know how this part work exactly, I just copied it from the "validateOne" implementation

Comment (pyphilia) [reply src/services/authorization.ts:60]: In fact that might be true, but rather:
```suggestion
 const isValid = highest && PermissionLevelCompare.gte(permission, highest);
```

**Pull request merged**